Sources/StreamChat/Database/DTOs/MessageDTO.swift

```diff
@@ -23,18 +23,24 @@ class MessageDTO: NSManagedObject {
     @NSManaged var extraData: Data
     @NSManaged var isSilent: Bool
     @NSManaged var reactionScores: [String: Int]
+    @NSManaged var reactionCounts: [String: Int]
     
     @NSManaged var user: UserDTO
     @NSManaged var mentionedUsers: Set<UserDTO>
     @NSManaged var threadParticipants: NSOrderedSet
     @NSManaged var channel: ChannelDTO?
     @NSManaged var replies: Set<MessageDTO>
     @NSManaged var flaggedBy: CurrentUserDTO?
-    @NSManaged var reactions: Set<MessageReactionDTO>
     @NSManaged var attachments: Set<AttachmentDTO>
     @NSManaged var quotedMessage: MessageDTO?
     @NSManaged var searches: Set<MessageSearchQueryDTO>
 
+    // We separate reactions from own reactions so that when paginating reactions,
+    // the own reactions don't break the pagination logic. This is why ownReactions
+    // are stored in a different property and as JSON Data.
+    @NSManaged var reactions: Set<MessageReactionDTO>
+    @NSManaged var ownReactions: Data
```

Comment:
Ok, it seems I understand why we have to store our own reactions separately 👍 **All** own reactions are included in the message payload but it's a subset of all reactions and they can be spread around the list. And to not mess up the list of reactions locally and follow the backend's pagination we store them in 2 different fields.

But why did we store own reactions as data and not as a relation? Is it some issue with the fetch request predicate?

Reply:
this is actually quite tricky :)

- with this approach we store reactions as core data objects *and* as JSON encoded data (easy to get things wrong)
- stashing latest reactions and own reactions like we do on `main` breaks pagination and probably other stuff

**Alternative Idea**

Store `ownReactions` as a list of reaction ids (Set<String>), this way we can avoid adding another serialization method (JSON) to the mix. Message.ownReactions is enriched at read time (get all reaction where ID IN ?) and the opposite is done at write time (Set<String> makes this trivial) 

Using a separate relationship also work but it is definetely more work and adds no real benefits

Reply:
@evsaev  If we store the own reactions normally in the DB, then the controller won't work because the fetch request gets all the reactions from the DB, it doesn't know which one belongs to own_reactions or not

Reply:
@tbarbugli The only issue I see right now with this approach is that if we change the Reactions structure, the DB migration won't affect the stored JSON... but I guess this won't happen any time soon.

On the other hand, changing the DB approach right now is a bit risky, without planning things properly.

Reply:
Yes, that is possible 👍 Let's keep a `UIViewController` as it is now